Add Check Support for Java 21 Unnamed Variables & Patterns Syntax: AnnotationOnSameLineCheck

child of https://github.com/checkstyle/checkstyle/issues/14942 
Check documentation: https://checkstyle.org/checks/annotation/annotationonsameline.html#AnnotationOnSameLine

```
PS D:\CS\test> javac src/Test.java
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
             <module name="AnnotationOnSameLine">
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java  
record ColoredPoint(int p, int x, String c) { }
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }
public class Test {
    void test(Object obj) {
        @Deprecated
        int _ = 0;

        if (obj instanceof ColoredPoint(@Deprecated // expected violation
                                          int _, _,_)){
        }
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:5:9: Annotation 'Deprecated' should be on the same line with its target. [AnnotationOnSameLine]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```
---

This is not very specific to unnamed patterns. but for patterns in general.
I expect a violation on the annotations on the pattern variable. The annotation should be on the same line as the pattern. We should add `PATTERN_VARIABLE_DEF` in the tokens

```
PS D:\CS\test> cat src/Test.java                                                
record ColoredPoint(int p, int x, String c) { } 
record Rectangle(ColoredPoint upperLeft, ColoredPoint lowerRight) { }  
public class Test {
    void test(Object obj) {
        if (obj instanceof ColoredPoint(@Deprecated // should this be violation ?
                                          int _, _,_)){
        }
        @Deprecated
        int _ = 0;
    }
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:8:9: Annotation 'Deprecated' should be on the same line with its target. [AnnotationOnSameLine]
Audit done.
Checkstyle ends with 1 errors.
PS D:\CS\test> 

```

---

I guess this should be ok this is all technically in the same `if `condition I think there is no point in requiring it to be on the same line

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationOnSameLineCheckTest.java
@@ -149,4 +149,14 @@ public class AnnotationOnSameLineCheckTest extends AbstractModuleTestSupport {
             expected);
     }
 
+    @Test
+    public void testAnnotationOnSameLinePatternVariables() throws Exception {
+        final String[] expected = {
+            "37:8: " + getCheckMessage(MSG_KEY_ANNOTATION_ON_SAME_LINE, "Deprecated"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputAnnotationOnSameLinePatternVariables.java"),
+            expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/annotationonsameline/InputAnnotationOnSameLinePatternVariables.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/annotation/annotationonsameline/InputAnnotationOnSameLinePatternVariables.java
@@ -1,0 +1,66 @@
+/*
+AnnotationOnSameLine
+tokens = (default)CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, \
+         VARIABLE_DEF, RECORD_DEF, COMPACT_CTOR_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.annotation.annotationonsameline;
+
+public class InputAnnotationOnSameLinePatternVariables {
+    void test(Object o) {
+        if (o instanceof ColoredPoint(@Deprecated
+                         int x,
+                         @Deprecated
+                         int y,
+                         @Deprecated
+                         String color)) { }
+
+       if (o instanceof ColoredPoint(@Deprecated
+                         int _,
+                         @Deprecated
+                         int _,
+                         @Deprecated
+                         String _)) { }
+
+       if (o instanceof @Deprecated
+                        String s) { }
+
+       if (o instanceof @Deprecated
+                        String _) { }
+
+       @Deprecated int _ = 0;
+
+       // violation below, 'Annotation 'Deprecated' should be on the same line with its target.'
+       @Deprecated
+       int _ = 0;
+    }
+
+    void test2(Object o) {
+        switch (o) {
+            case ColoredPoint(@Deprecated
+                              int x,
+                              @Deprecated
+                              int y,
+                              @Deprecated
+                              String color) when x >= 0 -> { }
+
+            case ColoredPoint(@Deprecated
+                              int x,
+                              @Deprecated
+                              int _,
+                              @Deprecated
+                              String color) -> { }
+            default -> {}
+        }
+    }
+
+    record ColoredPoint(@Deprecated
+                        int x,
+                        @Deprecated
+                        int y,
+                        @Deprecated
+                        String color) { }
+}

